src/go/goblins/docker/docker.go

```diff
@@ -0,0 +1,209 @@
+	default:
+		return errors.New("unknown driver")
+	}
+	if err != nil {
+		return err
+	}
+
+	for _, c := range commands {
+		fmt.Printf("Will now run the following command: %s ", c.Path)
+		for _, a := range c.Args {
+			fmt.Printf(" %s ", a)
+		}
+		fmt.Print("\n")
+		stdoutStderr, err := c.CombinedOutput()
+		fmt.Printf("%s\n", stdoutStderr)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	return nil
+}
+
+type storageInfo struct {
+	source      string
+	destination string
+	name        string
+}
+
+func aufsMountCommands(mountDir string, rootDir string, mountID string, dockerDir string,
+	storageMethod string, dockerVersion int, volumes map[string]string, mountPoints []storageInfo) ([]*exec.Cmd, error) {
+
+	if _, err := os.Stat(aufsPath); os.IsNotExist(err) {
+		return nil, errors.New("cannot find /sbin/mount.aufs. Maybe install aufs-tools?")
+	}
+	layersPath := filepath.Join(dockerDir, storageMethod, "layers", mountID)
+	containerID := filepath.Join(dockerDir, storageMethod, "diff", mountID)
+	if dockerVersion == 1 { // Old-style layers path.
+		layersPath = filepath.Join(dockerDir, storageMethod, "layers", containerID)
+	}
+
+	var commands []*exec.Cmd
+	mountpointPath := filepath.Join(dockerDir, storageMethod, "diff", containerID)
+	// Mount the base.
+	opts := fmt.Sprintf("-oro,br%s=ro+wh", mountpointPath)
+	args := []string{"-taufs", opts, "none", mountDir}
+	commands = append(commands, exec.Command("mount", args...))
+	// Now mount all layers above.
+	files, err := ioutil.ReadDir(layersPath)
+	if err != nil {
+		return nil, fmt.Errorf("Error while building layers mount command on layerspath %s: %v", layersPath, err)
+	}
+	for _, f := range files {
+		path := filepath.Join(dockerDir, storageMethod, "diff", f.Name())
+		opts := fmt.Sprintf("-oro,remount,append:%s=ro+wh", path)
+		args := []string{"-taufs", opts, "none", mountDir}
+		commands = append(commands, exec.Command("mount", args...))
+	}
+
+	// Now mount volumes.
+	switch dockerVersion {
+	case 1:
+		return append(commands, mountVolumesDocker1(mountDir, rootDir, volumes)...), nil
+	case 2:
+		return append(commands, mountVolumesDocker2(mountDir, mountPoints)...), nil
+	default:
+		return nil, fmt.Errorf("Unsupported docker version: %d", dockerVersion)
+	}
+}
+
+func mountVolumesDocker1(mountDir string, rootDir string, volumes map[string]string) []*exec.Cmd {
+	commands := make([]*exec.Cmd, 0)
+	for mountpoint, storage := range volumes {
+		storageIhp := strings.TrimLeft(mountpoint, string(os.PathSeparator))
+		mountpointIhp := strings.TrimLeft(storage, string(os.PathSeparator))
+		storagePath := filepath.Join(rootDir, storageIhp)
+		volumeMountpoint := filepath.Join(mountDir, mountpointIhp)
+		args := []string{"--bind", "-oro", storagePath, volumeMountpoint}
+		commands = append(commands, exec.Command("mount", args...))
+	}
+	return commands
+}
+
+func mountVolumesDocker2(mountDir string, mountPoints []storageInfo) []*exec.Cmd {
+	commands := make([]*exec.Cmd, 0)
```

Comment:
`var commands []*exec.Cmd`

The zero is not useful and you just append to the slice, so either size it correctly or let `append` allocate it for you. (`append` can work on nil slices)

Reply:
Fair. Let me fix it and the other instance.